crash report

```
type: crash
osVersion: google/coral/coral:13/TP1A.221005.002/2022101200:user/release-keys
package: com.google.android.gms:222615044
process: com.google.android.gms.persistent

java.lang.RuntimeException: Rethrown by StandaloneSchedulingEngine
	at beyi.a(:com.google.android.gms@222615044@22.26.15 (190400-461192076):12)
	at bevf.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):0)
	at xot.c(:com.google.android.gms@222615044@22.26.15 (190400-461192076):6)
	at xot.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):7)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at xty.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):0)
	at java.lang.Thread.run(Thread.java:1012)
	Suppressed: cfvr: 
		at tk_trace.ScheduledTaskImpl(Unknown Source:0)
		at tk_trace.stats-PlatformStatsCollectorChimeraService_onRunTaskAsync(Unknown Source:0)
		at tk_trace.scheduler-ServiceConnection_com.google.android.gms.stats.PlatformStatsCollectorService_onServiceConnected(Unknown Source:0)
		at tk_trace.checkQueueEagerly(Unknown Source:0)
		at tk_trace.trigger_EXECUTION_CALLBACK(Unknown Source:0)
		at tk_trace.scheduler-StandaloneSchedulingEngine_handleClientCallback(Unknown Source:0)
		at tk_trace.ScheduledTaskImpl(Unknown Source:0)
		at tk_trace.fitness-SyncGcmTaskChimeraService_onRunTaskAsync(Unknown Source:0)
		at tk_trace.scheduler-ServiceConnection_com.google.android.gms.fitness.sync.SyncGcmTaskService_onServiceConnected(Unknown Source:0)
		at tk_trace.checkQueueEagerly(Unknown Source:0)
		at tk_trace.trigger_EXECUTION_CALLBACK(Unknown Source:0)
		at tk_trace.scheduler-StandaloneSchedulingEngine_handleClientCallback(Unknown Source:0)
		at tk_trace.ScheduledTaskImpl(Unknown Source:0)
		at tk_trace.clearcut-QosUploaderChimeraService_onRunTaskAsync(Unknown Source:0)
		at tk_trace.scheduler-ServiceConnection_com.google.android.gms.clearcut.uploader.QosUploaderService_onServiceConnected(Unknown Source:0)
		at tk_trace.checkQueueEagerly(Unknown Source:0)
		at tk_trace.trigger_EXECUTION_CALLBACK(Unknown Source:0)
		at tk_trace.scheduler-StandaloneSchedulingEngine_handleClientCallback(Unknown Source:0)
		at tk_trace.ScheduledTaskImpl(Unknown Source:0)
		at tk_trace.tapandpay-TapAndPayGcmTaskChimeraService_onRunTaskAsync(Unknown Source:0)
		at tk_trace.scheduler-ServiceConnection_com.google.android.gms.tapandpay.gcmtask.TapAndPayGcmTaskService_onServiceConnected(Unknown Source:0)
		at tk_trace.checkQueueLazily(Unknown Source:0)
		at tk_trace.scheduler-StandaloneSchedulingEngine_trigger_EVENT_CODE_DOZE_LIGHT_MODE_CHANGED(Unknown Source:0)
		at tk_trace.scheduler-DeviceStateReceiver_onReceive(Unknown Source:0)
		at tk_trace.GmsCoreLogger.get(Unknown Source:0)
		at tk_trace.Missing root trace(Unknown Source:0)
Caused by: java.lang.RuntimeException: java.lang.NullPointerException
	at anyg.apply(:com.google.android.gms@222615044@22.26.15 (190400-461192076):2)
	at clmg.c(:com.google.android.gms@222615044@22.26.15 (190400-461192076):2)
	at clmh.run(:com.google.android.gms@222615044@22.26.15 (190400-461192076):11)
	... 6 more
Caused by: java.lang.NullPointerException
	at chma.a(:com.google.android.gms@222615044@22.26.15 (190400-461192076):0)
	at binf.a(:com.google.android.gms@222615044@22.26.15 (190400-461192076):2)
	at bine.c(:com.google.android.gms@222615044@22.26.15 (190400-461192076):55)
	at bilm.e(:com.google.android.gms@222615044@22.26.15 (190400-461192076):7)
	at bine.j(:com.google.android.gms@222615044@22.26.15 (190400-461192076):8)
	at bine.b(:com.google.android.gms@222615044@22.26.15 (190400-461192076):26)
	at bine.a(:com.google.android.gms@222615044@22.26.15 (190400-461192076):6)
	at com.google.android.gms.stats.PlatformStatsCollectorChimeraService.a(:com.google.android.gms@222615044@22.26.15 (190400-461192076):27)
	at anyd.call(:com.google.android.gms@222615044@22.26.15 (190400-461192076):4)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 6 more

```

--- a/core/java/android/app/SharedPreferencesImpl.java
+++ b/core/java/android/app/SharedPreferencesImpl.java
@@ -17,6 +17,7 @@
 package android.app;
 
 import android.annotation.Nullable;
+import android.app.compat.gms.GmsCompat;
 import android.compat.Compatibility;
 import android.compat.annotation.ChangeId;
 import android.compat.annotation.EnabledSince;
@@ -32,6 +33,7 @@ import android.system.StructTimespec;
 import android.util.Log;
 
 import com.android.internal.annotations.GuardedBy;
+import com.android.internal.gmscompat.GmsHooks;
 import com.android.internal.util.ExponentiallyBucketedHistogram;
 import com.android.internal.util.XmlUtils;
 
@@ -287,11 +289,24 @@ final class SharedPreferencesImpl implements SharedPreferences {
 
     @Override
     public Map<String, ?> getAll() {
+        HashMap<String, Object> res;
         synchronized (mLock) {
             awaitLoadedLocked();
             //noinspection unchecked
-            return new HashMap<String, Object>(mMap);
+            res = new HashMap<String, Object>(mMap);
         }
+
+        if (GmsCompat.isEnabled()) {
+            String fileName = mFile.getName();
+            String suffix = ".xml";
+            if (fileName.endsWith(suffix)) {
+                int endIndex = fileName.length() - suffix.length();
+                String name = fileName.substring(0, endIndex);
+                GmsHooks.maybeModifySharedPreferencesValues(name, res);
+            }
+        }
+
+        return res;
     }
 
     @Override

--- a/core/java/android/content/pm/GosPackageState.java
+++ b/core/java/android/content/pm/GosPackageState.java
@@ -50,6 +50,7 @@ public final class GosPackageState implements Parcelable {
     public static final int FLAG_ALLOW_ACCESS_TO_OBB_DIRECTORY = 1 << 1;
     public static final int FLAG_DISABLE_HARDENED_MALLOC = 1 << 2;
     public static final int FLAG_ENABLE_COMPAT_VA_39_BIT = 1 << 3;
+    // 1 << 4 was used by a now-removed feature, do not reuse it
 
     // to distinguish between the case when no dflags are set and the case when dflags weren't calculated yet
     public static final int DFLAGS_SET = 1;

--- a/core/java/com/android/internal/gmscompat/GmsHooks.java
+++ b/core/java/com/android/internal/gmscompat/GmsHooks.java
@@ -37,9 +37,7 @@ import android.content.Intent;
 import android.content.pm.ApplicationInfo;
 import android.content.pm.PackageManager;
 import android.database.Cursor;
-import android.database.CursorWrapper;
 import android.database.MatrixCursor;
-import android.database.sqlite.SQLiteDatabase;
 import android.database.sqlite.SQLiteOpenHelper;
 import android.net.Uri;
 import android.os.Build;
@@ -63,6 +61,7 @@ import com.android.internal.gmscompat.flags.GmsFlag;
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.List;
 import java.util.function.Consumer;
 
@@ -427,6 +426,19 @@ public final class GmsHooks {
         }
 
         return result;
+    }
+
+    // SharedPreferencesImpl#getAll
+    public static void maybeModifySharedPreferencesValues(String name, HashMap<String, Object> map) {
+        // some PhenotypeFlags are stored in SharedPreferences instead of phenotype.db database
+        ArrayMap<String, GmsFlag> flags = GmsHooks.config().flags.get(name);
+        if (flags == null) {
+            return;
+        }
+
+        for (GmsFlag f : flags.values()) {
+            f.applyToPhenotypeMap(map);
+        }
     }
 
     // Instrumentation#execStartActivity(Context, IBinder, IBinder, Activity, Intent, int, Bundle)

--- a/core/java/com/android/internal/gmscompat/flags/GmsFlag.java
+++ b/core/java/com/android/internal/gmscompat/flags/GmsFlag.java
@@ -15,6 +15,7 @@ import android.util.Log;
 
 import com.android.internal.gmscompat.GmsInfo;
 
+import java.util.Map;
 import java.util.function.Supplier;
 
 public class GmsFlag implements Parcelable {
@@ -103,7 +104,7 @@ public class GmsFlag implements Parcelable {
 
     private static final int PHENOTYPE_BASE64_FLAGS = Base64.NO_PADDING | Base64.NO_WRAP;
 
-    public void applyToPhenotypeMap(ArrayMap<String, String> map) {
+    public void applyToPhenotypeMap(Map map) {
         if (!shouldOverride()) {
             return;
         }
@@ -134,7 +135,7 @@ public class GmsFlag implements Parcelable {
                 s = Double.toString(floatArg);
                 break;
             case TYPE_STRING:
-                s = stringVal(map.get(name));
+                s = stringVal(String.valueOf(map.get(name)));
                 break;
             case TYPE_BYTES:
                 s = Base64.encodeToString(bytesArg, PHENOTYPE_BASE64_FLAGS);

--- a/services/core/java/com/android/server/pm/permission/SpecialRuntimePermUtils.java
+++ b/services/core/java/com/android/server/pm/permission/SpecialRuntimePermUtils.java
@@ -29,6 +29,7 @@ import android.util.SparseArray;
 
 import com.android.server.LocalServices;
 import com.android.server.ext.SystemServerExt;
+import com.android.server.pm.UserManagerInternal;
 import com.android.server.pm.parsing.pkg.AndroidPackage;
 import com.android.server.pm.pkg.component.ParsedUsesPermission;
 
@@ -63,13 +64,17 @@ public class SpecialRuntimePermUtils {
                 return false;
             }
 
+            var um = LocalServices.getService(UserManagerInternal.class);
+            // use parent profile settings for work profile
+            int userIdForSettings = um.getProfileParentId(userId);
+
             Context ctx = SystemServerExt.get().context;
             var cr = ctx.getContentResolver();
             var key = Settings.Secure.AUTO_GRANT_OTHER_SENSORS_PERMISSION;
             int def = Settings.Secure.AUTO_GRANT_OTHER_SENSORS_PERMISSION_DEFAULT;
 
             try {
-                return Settings.Secure.getIntForUser(cr, key, def, userId) == 1;
+                return Settings.Secure.getIntForUser(cr, key, def, userIdForSettings) == 1;
             } catch (Exception e) {
                 // a failsafe: should never happen
                 Slog.d(TAG, "", e);

